cluster-health-check/cluster-health.sh

```diff
@@ -0,0 +1,519 @@
+SPACE=2048
+setup() {
+  TMPDIR=$(mktemp -d $MKTEMP_BASEDIR)
+  techo "Created ${TMPDIR}"
+}
+disk-space() {
+  AVAILABLE=$(df -m ${TMPDIR} | tail -n 1 | awk '{ print $4 }')
+  if [ "${AVAILABLE}" -lt "${SPACE}" ]
+    then
+      techo "${AVAILABLE} MB space free, minimum needed is ${SPACE} MB."
+      DISK_FULL=1
+  fi
+}
+verify-access() {
+  techo "Verifying cluster access"
+  if [[ ! -z $OVERRIDE_KUBECONFIG ]];
+  then
+    ## Just use the kubeconfig that was set by the user
+    KUBECTL_CMD="kubectl --kubeconfig $OVERRIDE_KUBECONFIG"
+  elif [[ ! -z $KUBECONFIG ]];
+  then
+    KUBECTL_CMD="kubectl"
+  elif [[ ! -z $KUBERNETES_PORT ]];
+  then
+    ## We are inside the k8s cluster or we're using the local kubeconfig
+    RANCHER_POD=$(kubectl -n cattle-system get pods -l app=rancher --no-headers -o custom-columns=id:metadata.name | head -n1)
+    KUBECTL_CMD="kubectl -n cattle-system exec -c rancher ${RANCHER_POD} -- kubectl"
+  elif $(command -v k3s >/dev/null 2>&1)
+  then
+    ## We are on k3s node
+    KUBECTL_CMD="k3s kubectl"
+  elif $(command -v docker >/dev/null 2>&1)
+  then
+    DOCKER_ID=$(docker ps | grep "k8s_rancher_rancher" | cut -d' ' -f1 | head -1)
+    KUBECTL_CMD="docker exec ${DOCKER_ID} kubectl"
+  else
+    ## Giving up
+    techo "Could not find a kubeconfig"
+    exit 2
+  fi
+  if ! ${KUBECTL_CMD} cluster-info >/dev/null 2>&1
+  then
+    techo "Can not access cluster"
+    exit 1
+  else
+    techo "Cluster access has been verified"
+  fi
+  if [[ ! -z $DEBUG ]]
+  then
+    KUBECTL_CMD="${KUBECTL_CMD} --v=6"
+  fi
+}
+cluster-info() {
+  techo "Collecting information about the cluster..."
+  mkdir -p $TMPDIR/clusterinfo
+  decho "cluster-info"
+  ${KUBECTL_CMD} cluster-info > $TMPDIR/clusterinfo/cluster-info 2>&1
+  ${KUBECTL_CMD} cluster-info dump > $TMPDIR/clusterinfo/cluster-info-dump 2>&1
+  decho "get nodes"
+  ${KUBECTL_CMD} get nodes -o wide > $TMPDIR/clusterinfo/get-node.wide 2>&1
+  ${KUBECTL_CMD} get nodes -o yaml > $TMPDIR/clusterinfo/get-node-yaml 2>&1
+}
+detect-provider() {
+  techo "Detecting the cluster provider..."
+  mkdir -p $TMPDIR/clusterprovider
+  decho "cluster-info"
+  clusterprovider=`${KUBECTL_CMD} get cluster -o json | jq -r '.items[].status.provider'`
+  echo $clusterprovider > $TMPDIR/clusterprovider/provider
+}
+nodes() {
+  techo "Collecting information about the nodes..."
+  mkdir -p $TMPDIR/nodes/
+  ${KUBECTL_CMD} get nodes -o wide > $TMPDIR/nodes/get-nodes.wide
+  ${KUBECTL_CMD} get nodes -o yaml > $TMPDIR/nodes/get-nodes-yaml
+  ${KUBECTL_CMD} top nodes > $TMPDIR/nodes/top-node
+  decho "Describing nodes..."
+  mkdir -p $TMPDIR/nodes/describe
+  for NODE in `${KUBECTL_CMD} get node -o NAME | awk -F '/' '{print $2}'`
+  do
+    ${KUBECTL_CMD} describe node $NODE > $TMPDIR/nodes/describe/$NODE
+  done
+  techo "Gathering node counts..."
+  decho "Total number of nodes..."
+  NumberOfNodes=`${KUBECTL_CMD} get nodes --no-headers | wc -l`
```

Comment:
We should run any `kubectl` command only once in the script. Rerunnning the same command over and over increases the overall runtime of the script. Ideally, if you can grab the info using `-o json` and redirect it to a file, you can then use `jq` commands to gather the necessary information. Also, we should split the logic into two different phases. First one does just collection of the information in json format, the second format would summarize the necessary information.. This would help reduce the overall bundle size too.

Reply:
I consolidated as much of the kubectl commands as I could. I did run into issues with the role selector for the node section as it looks like jq doesn't like the '/' in the role name.